File: supporthub-sdks/react-native/src/SupportHub.ts
import { SupportHubConfig, UserIdentity, ProjectSettings, SupportHubError } from './models/types';
import { ApiClient } from './api/client';
import { SupportHubStorage } from './storage/storage';
import { DEFAULT_CONFIG } from './config';

/**
 * Main SupportHub SDK class (singleton).
 *
 * Usage:
 * ```ts
 * await SupportHub.initialize({ projectId: '...', apiKey: '...' });
 * await SupportHub.identify({ externalId: 'user-123', name: 'John' });
 * ```
 */
export class SupportHub {
  private static instance: SupportHub | null = null;

  private config: SupportHubConfig;
  private apiClient: ApiClient;
  private currentUser: UserIdentity | null = null;
  private settings: ProjectSettings | null = null;
  private pollIntervalMs: number;
  private unreadPollTimer: ReturnType<typeof setInterval> | null = null;
  private unreadListeners: Set<(count: number) => void> = new Set();
  private lastUnreadCount: number = 0;

  // ─── Constructor ────────────────────────────────────────────────────

  private constructor(config: SupportHubConfig) {
    this.config = config;
    this.apiClient = new ApiClient(config);
    this.pollIntervalMs = config.pollIntervalMs || DEFAULT_CONFIG.pollIntervalMs;
  }

  // ─── Initialization ─────────────────────────────────────────────────

  /**
   * Initialize the SupportHub SDK.
   * Must be called once before any other SDK method.
   * Fetches and caches project settings.
   */
  static async initialize(config: SupportHubConfig): Promise<void> {
    if (!config.projectId || !config.apiKey) {
      throw new SupportHubError(
        'projectId and apiKey are required',
        0,
        'INVALID_CONFIG',
      );
    }

    SupportHub.instance = new SupportHub(config);

    // Pre-fetch settings (non-blocking — cache for later use)
    try {
      await SupportHub.instance.fetchSettings();
    } catch {
      // Settings fetch failure is non-fatal during init
    }

    // Restore previous user session from storage
    try {
      const storedId = await SupportHubStorage.getExternalId();
      if (storedId) {
        SupportHub.instance.currentUser = { externalId: storedId };
        SupportHub.instance.startUnreadPolling();
      }
    } catch {
      // Storage failure is non-fatal
    }
  }

  /**
   * Get the singleton instance.
   * @throws SupportHubError if not initialized.
   */
  static getInstance(): SupportHub {
    if (!SupportHub.instance) {
      throw new SupportHubError(
        'SupportHub not initialized. Call SupportHub.initialize() first.',
        0,
        'NOT_INITIALIZED',
      );
    }
    return SupportHub.instance;
  }

  /** Get the underlying API client. Used internally by hooks and screens. */
  getApiClient(): ApiClient {
    return this.apiClient;
  }

  /** Get the configured polling interval in milliseconds. */
  getPollIntervalMs(): number {
    return this.pollIntervalMs;
  }

  // ─── User Identity ──────────────────────────────────────────────────

  /**
   * Identify a known user. Sends identity to the API and starts
   * polling for unread messages.
   */
  static async identify(identity: UserIdentity): Promise<void> {
    const hub = SupportHub.getInstance();

    if (!identity.externalId) {
      throw new SupportHubError(
        'externalId is required for identify()',
        0,
        'INVALID_IDENTITY',
      );
    }

    // Send identity to API
    try {
      await hub.apiClient.identifyUser(identity);
    } catch (err) {
      // Log but don't block — user can still use cached data
      console.warn('[SupportHub] Failed to identify user:', err);
    }

    hub.currentUser = identity;

    // Persist for session restore
    await SupportHubStorage.setExternalId(identity.externalId);
    await SupportHubStorage.setUserIdentity(identity as Record<string, any>);

    // Start polling for unread messages
    hub.startUnreadPolling();
  }

  /**
   * Identify an anonymous user.
   * Generates a persistent anonymous ID stored in AsyncStorage.
   */
  static async identifyAnonymous(): Promise<void> {
    const anonymousId = await SupportHubStorage.getAnonymousId();

    await SupportHub.identify({
      externalId: anonymousId,
      identifierType: 'anonymous',
    });
  }

  /** Get the currently identified user, or null if not identified. */
  static getCurrentUser(): UserIdentity | null {
    if (!SupportHub.instance) return null;
    return SupportHub.instance.currentUser;
  }

  // ─── Unread Count ───────────────────────────────────────────────────

  /** Fetch the current unread message count. */
  static async getUnreadCount(): Promise<number> {
    const hub = SupportHub.getInstance();

    if (!hub.currentUser) {
      return 0;
    }

    try {
      const count = await hub.apiClient.getUnreadCount(hub.currentUser.externalId);
      hub.lastUnreadCount = count;
      return count;
    } catch {
      return hub.lastUnreadCount;
    }
  }

  /**
   * Subscribe to unread count changes.
   * The listener is called whenever the polled unread count changes.
   * @returns An unsubscribe function.
   */
  static onUnreadCountChanged(listener: (count: number) => void): () => void {
    const hub = SupportHub.getInstance();
    hub.unreadListeners.add(listener);

    // Immediately notify with current value
    listener(hub.lastUnreadCount);

    return () => {
      hub.unreadListeners.delete(listener);
    };
  }

  // ─── Device Registration ────────────────────────────────────────────

  /** Register a device push notification token. */
  static async registerDeviceToken(
    token: string,
    platform: 'android' | 'ios',
  ): Promise<void> {
    const hub = SupportHub.getInstance();

    if (!hub.currentUser) {
      throw new SupportHubError(
        'User must be identified before registering a device token.',
        0,
        'NOT_IDENTIFIED',
      );
    }

    await hub.apiClient.registerDevice(hub.currentUser.externalId, token, platform);
  }

  /** Unregister a device push notification token. */
  static async unregisterDeviceToken(token: string): Promise<void> {
    const hub = SupportHub.getInstance();
    await hub.apiClient.unregisterDevice(token);
  }

  // ─── Settings ───────────────────────────────────────────────────────

  /**
   * Get project settings (branding, auto-response config).
   * Returns cached settings if available, otherwise fetches from API.
   */
  static async getSettings(): Promise<ProjectSettings> {
    const hub = SupportHub.getInstance();
    if (hub.settings) {
      return hub.settings;
    }
    return hub.fetchSettings();
  }

  /** Fetch settings from API and cache locally. */
  private async fetchSettings(): Promise<ProjectSettings> {
    try {
      this.settings = await this.apiClient.getSettings();
      await SupportHubStorage.cacheSettings(this.settings as Record<string, any>);
      return this.settings;
    } catch {
      // Try loading from cache
      const cached = await SupportHubStorage.getCachedSettings();
      if (cached) {
        this.settings = cached as unknown as ProjectSettings;
        return this.settings;
      }
      // Return defaults
      this.settings = {
        widgetTitle: 'Support',
        welcomeMessage: 'Hi! How can we help you today?',
        brandingPrimaryColor: '#6366F1',
        brandingBgColor: '#0F172A',
        showReadIndicators: true,
        autoResponseEnabled: false,
        autoResponseMessage: '',
      };
      return this.settings;
    }
  }

  // ─── Logout ─────────────────────────────────────────────────────────

  /**
   * Logout the current user.
   * Stops polling, clears storage, and resets all state.
   */
  static async logout(): Promise<void> {
    if (!SupportHub.instance) return;

    const hub = SupportHub.instance;
    hub.stopUnreadPolling();
    hub.currentUser = null;
    hub.settings = null;
    hub.lastUnreadCount = 0;
    hub.unreadListeners.clear();

    await SupportHubStorage.clear();

    SupportHub.instance = null;
  }

  // ─── Polling ────────────────────────────────────────────────────────

  /** Start polling for unread count at the configured interval. */
  private startUnreadPolling(): void {
    // Don't start duplicate timers
    this.stopUnreadPolling();

    if (!this.currentUser) return;

    this.unreadPollTimer = setInterval(async () => {
      if (!this.currentUser) {
        this.stopUnreadPolling();
        return;
      }

      try {
        const count = await this.apiClient.getUnreadCount(this.currentUser.externalId);
        if (count !== this.lastUnreadCount) {
          this.lastUnreadCount = count;
          this.notifyUnreadListeners(count);
        }
      } catch {
        // Silently ignore polling errors
      }
    }, this.pollIntervalMs);
  }

  /** Stop the unread count polling timer. */
  private stopUnreadPolling(): void {
    if (this.unreadPollTimer) {
      clearInterval(this.unreadPollTimer);
      this.unreadPollTimer = null;
    }
  }

  /** Notify all registered unread count listeners. */
  private notifyUnreadListeners(count: number): void {
    this.unreadListeners.forEach((listener) => {
      try {
        listener(count);
      } catch {
        // Don't let a bad listener break others
      }
    });
  }
}

File: supporthub-sdks/flutter/lib/src/supporthub.dart
import 'dart:async';

import 'package:flutter/material.dart';

import 'api/api_client.dart';
import 'config.dart';
import 'models/project_settings.dart';
import 'models/user_identity.dart';
import 'screens/messaging_screen.dart';
import 'storage/local_storage.dart';

/// The main entry-point for the SupportHub SDK.
///
/// Call [SupportHub.initialize] once during app startup, then use
/// [SupportHub.identify] (or [identifyAnonymous]) and
/// [SupportHub.open] to present the messaging UI.
///
/// ```dart
/// await SupportHub.initialize(
///   projectId: 'my-project',
///   apiKey: 'sk_live_xxx',
/// );
///
/// await SupportHub.identify(externalId: 'user@example.com');
///
/// SupportHub.open(context);
/// ```
class SupportHub {
  // ── Singleton ─────────────────────────────────────────────────────

  static SupportHub? _instance;

  /// Returns the initialised SDK instance.
  ///
  /// Throws [StateError] if [initialize] has not been called.
  static SupportHub get instance {
    if (_instance == null) {
      throw StateError(
        'SupportHub has not been initialised. '
        'Call SupportHub.initialize() first.',
      );
    }
    return _instance!;
  }

  /// Whether the SDK has been initialised.
  static bool get isInitialized => _instance != null;

  // ── Fields ────────────────────────────────────────────────────────

  /// The active configuration.
  final SupportHubConfig config;

  /// The HTTP client used to communicate with the API.
  final ApiClient apiClient;

  /// Local persistence.
  final LocalStorage storage;

  /// The currently identified user, or `null`.
  UserIdentity? _currentUser;

  /// Cached project settings.
  ProjectSettings? _settings;

  /// Controller for the unread-count stream.
  StreamController<int>? _unreadController;

  /// Timer used for polling unread counts.
  Timer? _pollTimer;

  // ── Constructor (private) ─────────────────────────────────────────

  SupportHub._({
    required this.config,
    required this.apiClient,
    required this.storage,
  });

  // ── Public API ────────────────────────────────────────────────────

  /// Initialises the SDK. Must be called once before any other method.
  ///
  /// [projectId] and [apiKey] are required. [apiUrl] defaults to a
  /// placeholder Supabase URL — replace with your own.
  static Future<void> initialize({
    required String projectId,
    required String apiKey,
    String? apiUrl,
    Duration pollInterval = const Duration(seconds: 30),
  }) async {
    WidgetsFlutterBinding.ensureInitialized();

    final config = SupportHubConfig(
      projectId: projectId,
      apiKey: apiKey,
      apiUrl: apiUrl ?? 'https://$projectId.supabase.co/functions/v1/supporthub',
      pollInterval: pollInterval,
    );

    final client = ApiClient(config);
    final localStorage = LocalStorage();

    _instance = SupportHub._(
      config: config,
      apiClient: client,
      storage: localStorage,
    );

    // Pre-fetch settings in the background.
    unawaited(_instance!._loadSettings());
  }

  /// The currently identified user.
  static UserIdentity? get currentUser => instance._currentUser;

  /// Cached project settings.
  static ProjectSettings? get settings => instance._settings;

  // ── Identify ──────────────────────────────────────────────────────

  /// Identifies an authenticated user.
  ///
  /// Upserts the user in the SupportHub backend and caches the
  /// identity locally.
  static Future<void> identify({
    required String externalId,
    String identifierType = 'email',
    String? email,
    String? name,
    String? company,
    String? subscriptionTier,
    Map<String, dynamic>? metadata,
  }) async {
    final identity = UserIdentity(
      externalId: externalId,
      identifierType: identifierType,
      email: email,
      name: name,
      company: company,
      subscriptionTier: subscriptionTier,
      metadata: metadata,
    );

    try {
      await instance.apiClient.identifyUser(identity.toJson());
    } catch (e, stack) {
      print('[SupportHub SDK] Error identifying user: $e');
      print(stack);
      // Identification is best-effort — allow offline usage.
    }

    instance._currentUser = identity;
    await instance.storage.setExternalId(externalId);
  }

  /// Identifies an anonymous user using a persistent device ID.
  ///
  /// A UUID is generated on first call and reused across sessions.
  static Future<void> identifyAnonymous() async {
    final anonId = await instance.storage.getAnonymousId();

    final identity = UserIdentity(
      externalId: anonId,
      identifierType: 'device_id',
    );

    try {
      await instance.apiClient.identifyUser(identity.toJson());
    } catch (e, stack) {
      print('[SupportHub SDK] Error identifying anonymous user: $e');
      print(stack);
      // Best-effort.
    }

    instance._currentUser = identity;
    await instance.storage.setExternalId(anonId);
  }

  // ── Open UI ───────────────────────────────────────────────────────

  /// Opens the SupportHub messaging screen.
  ///
  /// Pushes a full-screen [MessagingScreen] onto the navigator.
  /// If no user has been identified, [identifyAnonymous] is called
  /// automatically.
  static void open(BuildContext context) {
    if (instance._currentUser == null) {
      identifyAnonymous().then((_) => _pushMessaging(context));
    } else {
      _pushMessaging(context);
    }
  }

  static void _pushMessaging(BuildContext context) {
    Navigator.of(context).push(
      MaterialPageRoute<void>(
        builder: (_) => const MessagingScreen(),
      ),
    );
  }

  // ── Unread Count ──────────────────────────────────────────────────

  /// Returns the current unread message count for the identified user.
  static Future<int> getUnreadCount() async {
    final userId = instance._currentUser?.externalId;
    if (userId == null) return 0;

    try {
      return await instance.apiClient.getUnreadCount(userId);
    } catch (_) {
      return 0;
    }
  }

  /// A broadcast stream that emits the unread count periodically.
  ///
  /// Polls every [SupportHubConfig.pollInterval] (default 30 s).
  /// Cancel the subscription when the widget is disposed to stop
  /// polling.
  static Stream<int> onUnreadCountChanged() {
    final hub = instance;

    hub._unreadController ??= StreamController<int>.broadcast(
      onListen: () => hub._startPolling(),
      onCancel: () => hub._stopPolling(),
    );

    return hub._unreadController!.stream;
  }

  void _startPolling() {
    _pollTimer?.cancel();

    // Emit immediately.
    _emitUnread();

    _pollTimer = Timer.periodic(config.pollInterval, (_) => _emitUnread());
  }

  void _stopPolling() {
    _pollTimer?.cancel();
    _pollTimer = null;
  }

  Future<void> _emitUnread() async {
    try {
      final count = await getUnreadCount();
      _unreadController?.add(count);
    } catch (_) {
      // Silently ignore polling errors.
    }
  }

  // ── Push Notifications ────────────────────────────────────────────

  /// Registers a device token for push notifications.
  ///
  /// [platform] should be `'android'`, `'ios'`, or `'web'`.
  static Future<void> registerDeviceToken({
    required String token,
    required String platform,
  }) async {
    final userId = instance._currentUser?.externalId;
    if (userId == null) return;

    await instance.apiClient.registerDevice(userId, token, platform);
  }

  /// Unregisters a previously-registered device token.
  static Future<void> unregisterDeviceToken(String token) async {
    await instance.apiClient.unregisterDevice(token);
  }

  // ── Logout ────────────────────────────────────────────────────────

  /// Clears the current user session and all persisted data.
  static Future<void> logout() async {
    instance._stopPolling();
    instance._unreadController?.close();
    instance._unreadController = null;
    instance._currentUser = null;
    instance._settings = null;
    await instance.storage.clearAll();
  }

  // ── Settings ──────────────────────────────────────────────────────

  /// Loads project settings from the API.
  ///
  /// Returns cached settings if available.
  static Future<ProjectSettings> loadSettings() async {
    if (instance._settings != null) return instance._settings!;
    return instance._loadSettings();
  }

  Future<ProjectSettings> _loadSettings() async {
    try {
      final json = await apiClient.getSettings();
      _settings = ProjectSettings.fromJson(json);
    } catch (e, stack) {
      print('[SupportHub SDK] Error loading settings: $e');
      print(stack);
      _settings = const ProjectSettings();
    }
    return _settings!;
  }
}
